// @sharpee/helpers/create-helpers.js
"use strict";
/**
 * createHelpers — factory that binds a world reference and returns
 * all entity builder functions.
 *
 * Public interface: createHelpers(world) => EntityHelpers
 *
 * Owner context: @sharpee/helpers (ADR-140)
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createHelpers = createHelpers;
const room_1 = require("./builders/room");
const object_1 = require("./builders/object");
const container_1 = require("./builders/container");
const actor_1 = require("./builders/actor");
const door_1 = require("./builders/door");
/**
 * Create a set of entity builder functions bound to a world reference.
 *
 * @param world - The IWorldModel to create entities in
 * @returns EntityHelpers with all builder factories
 */
function createHelpers(world) {
    return {
        room: (name) => new room_1.RoomBuilder(world, name),
        object: (name) => new object_1.ObjectBuilder(world, name),
        container: (name) => new container_1.ContainerBuilder(world, name),
        actor: (name) => new actor_1.ActorBuilder(world, name),
        door: (name) => new door_1.DoorBuilder(world, name),
    };
}
