// @sharpee/helpers/index.js
"use strict";
/**
 * @sharpee/helpers — Fluent entity builder helpers for the Sharpee IF platform.
 *
 * Public interface: Importing this module activates world.helpers() via
 * declaration merging on IWorldModel. Also exports createHelpers() and
 * EntityHelpers for direct use.
 *
 * Owner context: @sharpee/helpers (ADR-140)
 *
 * @example
 * ```typescript
 * import "./index.js";
 *
 * initializeWorld(world: WorldModel): void {
 *   const { room, object, actor } = world.helpers();
 *
 *   const kitchen = room('Kitchen')
 *     .description('A warm kitchen.')
 *     .build();
 * }
 * ```
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DoorBuilder = exports.ActorBuilder = exports.ContainerBuilder = exports.ObjectBuilder = exports.RoomBuilder = exports.createHelpers = void 0;
// Side-effect import: patches WorldModel.prototype.helpers
require("./augment");
// Named exports for direct use
var create_helpers_1 = require("./create-helpers");
Object.defineProperty(exports, "createHelpers", { enumerable: true, get: function () { return create_helpers_1.createHelpers; } });
var room_1 = require("./builders/room");
Object.defineProperty(exports, "RoomBuilder", { enumerable: true, get: function () { return room_1.RoomBuilder; } });
var object_1 = require("./builders/object");
Object.defineProperty(exports, "ObjectBuilder", { enumerable: true, get: function () { return object_1.ObjectBuilder; } });
var container_1 = require("./builders/container");
Object.defineProperty(exports, "ContainerBuilder", { enumerable: true, get: function () { return container_1.ContainerBuilder; } });
var actor_1 = require("./builders/actor");
Object.defineProperty(exports, "ActorBuilder", { enumerable: true, get: function () { return actor_1.ActorBuilder; } });
var door_1 = require("./builders/door");
Object.defineProperty(exports, "DoorBuilder", { enumerable: true, get: function () { return door_1.DoorBuilder; } });
