// @sharpee/helpers/builders/actor.js
"use strict";
/**
 * ActorBuilder — fluent builder for actor entities.
 *
 * Public interface: description, aliases, properName, inventory, in,
 * skipValidation, build.
 *
 * Owner context: @sharpee/helpers (ADR-140)
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActorBuilder = void 0;
const world_model_1 = require("../../world-model/index.js");
/**
 * Fluent builder for creating actor entities (players and NPCs).
 *
 * @example
 * ```typescript
 * const player = actor('yourself')
 *   .description('As good-looking as ever.')
 *   .aliases('self', 'me', 'myself')
 *   .properName()
 *   .inventory({ maxItems: 10 })
 *   .build();
 * ```
 */
class ActorBuilder {
    world;
    name;
    _description;
    _aliases;
    _properName = false;
    _inventory;
    _location;
    _skipValidation = false;
    _traits = [];
    constructor(world, name) {
        this.world = world;
        this.name = name;
    }
    /**
     * Set the actor description.
     *
     * @param desc - Actor description text
     * @returns this (for chaining)
     */
    description(desc) {
        this._description = desc;
        return this;
    }
    /**
     * Add name aliases.
     *
     * @param names - Alternative names
     * @returns this (for chaining)
     */
    aliases(...names) {
        this._aliases = names;
        return this;
    }
    /**
     * Mark as a proper name (no article).
     *
     * @returns this (for chaining)
     */
    properName() {
        this._properName = true;
        return this;
    }
    /**
     * Add inventory capacity (adds ContainerTrait).
     *
     * @param opts - Inventory options
     * @returns this (for chaining)
     */
    inventory(opts = {}) {
        this._inventory = opts;
        return this;
    }
    /**
     * Place the actor in a location.
     *
     * @param location - The entity to place this actor in
     * @returns this (for chaining)
     */
    in(location) {
        this._location = location;
        return this;
    }
    /**
     * Add a custom trait to the entity.
     *
     * @param trait - Any ITrait instance
     * @returns this (for chaining)
     */
    addTrait(trait) {
        this._traits.push(trait);
        return this;
    }
    /**
     * Bypass validation for placement.
     *
     * @returns this (for chaining)
     */
    skipValidation() {
        this._skipValidation = true;
        return this;
    }
    /**
     * Create the actor entity with configured traits.
     *
     * @returns The created IFEntity
     */
    build() {
        const entity = this.world.createEntity(this.name, 'actor');
        entity.add(new world_model_1.ActorTrait());
        entity.add(new world_model_1.IdentityTrait({
            name: this.name,
            description: this._description,
            aliases: this._aliases,
            properName: this._properName,
            article: this._properName ? '' : undefined,
        }));
        if (this._inventory) {
            entity.add(new world_model_1.ContainerTrait({
                capacity: this._inventory.maxItems ? { maxItems: this._inventory.maxItems } : undefined,
            }));
        }
        for (const trait of this._traits) {
            entity.add(trait);
        }
        if (this._location) {
            const mover = this._skipValidation
                ? new world_model_1.AuthorModel(this.world.getDataStore(), this.world)
                : this.world;
            mover.moveEntity(entity.id, this._location.id);
        }
        return entity;
    }
}
exports.ActorBuilder = ActorBuilder;
