// @sharpee/helpers/builders/container.js
"use strict";
/**
 * ContainerBuilder — fluent builder for container entities.
 *
 * Public interface: description, aliases, in, openable, lockable,
 * skipValidation, build.
 *
 * Owner context: @sharpee/helpers (ADR-140)
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContainerBuilder = void 0;
const world_model_1 = require("../../world-model/index.js");
/**
 * Fluent builder for creating container entities.
 *
 * @example
 * ```typescript
 * const chest = container('wooden chest')
 *   .description('A sturdy wooden chest.')
 *   .openable({ isOpen: false })
 *   .lockable({ isLocked: true, keyId: key.id })
 *   .in(treasureRoom)
 *   .build();
 * ```
 */
class ContainerBuilder {
    world;
    name;
    _description;
    _aliases;
    _location;
    _openable;
    _lockable;
    _skipValidation = false;
    _traits = [];
    constructor(world, name) {
        this.world = world;
        this.name = name;
    }
    /**
     * Set the container description.
     *
     * @param desc - Container description text
     * @returns this (for chaining)
     */
    description(desc) {
        this._description = desc;
        return this;
    }
    /**
     * Add name aliases.
     *
     * @param names - Alternative names
     * @returns this (for chaining)
     */
    aliases(...names) {
        this._aliases = names;
        return this;
    }
    /**
     * Place the container in a location.
     *
     * @param location - The entity to place this container in
     * @returns this (for chaining)
     */
    in(location) {
        this._location = location;
        return this;
    }
    /**
     * Make the container openable.
     *
     * @param opts - Openable options
     * @returns this (for chaining)
     */
    openable(opts = {}) {
        this._openable = opts;
        return this;
    }
    /**
     * Make the container lockable.
     *
     * @param opts - Lockable options
     * @returns this (for chaining)
     */
    lockable(opts = {}) {
        this._lockable = opts;
        return this;
    }
    /**
     * Add a custom trait to the entity.
     *
     * @param trait - Any ITrait instance
     * @returns this (for chaining)
     */
    addTrait(trait) {
        this._traits.push(trait);
        return this;
    }
    /**
     * Bypass validation for placement.
     *
     * @returns this (for chaining)
     */
    skipValidation() {
        this._skipValidation = true;
        return this;
    }
    /**
     * Create the container entity with configured traits.
     *
     * @returns The created IFEntity
     */
    build() {
        const entity = this.world.createEntity(this.name, 'object');
        entity.add(new world_model_1.IdentityTrait({
            name: this.name,
            description: this._description,
            aliases: this._aliases,
        }));
        entity.add(new world_model_1.ContainerTrait());
        if (this._openable) {
            entity.add(new world_model_1.OpenableTrait({ isOpen: this._openable.isOpen ?? true }));
        }
        if (this._lockable) {
            entity.add(new world_model_1.LockableTrait({
                isLocked: this._lockable.isLocked ?? false,
                ...(this._lockable.keyId ? { requiredKey: this._lockable.keyId } : {}),
            }));
        }
        for (const trait of this._traits) {
            entity.add(trait);
        }
        if (this._location) {
            const mover = this._skipValidation
                ? new world_model_1.AuthorModel(this.world.getDataStore(), this.world)
                : this.world;
            mover.moveEntity(entity.id, this._location.id);
        }
        return entity;
    }
}
exports.ContainerBuilder = ContainerBuilder;
