// @sharpee/helpers/builders/object.js
"use strict";
/**
 * ObjectBuilder — fluent builder for object entities.
 *
 * Public interface: description, aliases, in, scenery, lightSource,
 * skipValidation, build.
 *
 * Owner context: @sharpee/helpers (ADR-140)
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectBuilder = void 0;
const world_model_1 = require("../../world-model/index.js");
/**
 * Fluent builder for creating object entities.
 *
 * @example
 * ```typescript
 * const knife = object('bread knife')
 *   .description('A sharp bread knife.')
 *   .aliases('knife', 'blade')
 *   .in(kitchen)
 *   .build();
 * ```
 */
class ObjectBuilder {
    world;
    name;
    _description;
    _aliases;
    _location;
    _scenery = false;
    _lightSource;
    _skipValidation = false;
    _traits = [];
    constructor(world, name) {
        this.world = world;
        this.name = name;
    }
    /**
     * Set the object description.
     *
     * @param desc - Object description text
     * @returns this (for chaining)
     */
    description(desc) {
        this._description = desc;
        return this;
    }
    /**
     * Add name aliases for the object.
     *
     * @param names - Alternative names
     * @returns this (for chaining)
     */
    aliases(...names) {
        this._aliases = names;
        return this;
    }
    /**
     * Place the object in a location.
     *
     * @param location - The entity to place this object in
     * @returns this (for chaining)
     */
    in(location) {
        this._location = location;
        return this;
    }
    /**
     * Mark as scenery (non-portable).
     *
     * @returns this (for chaining)
     */
    scenery() {
        this._scenery = true;
        return this;
    }
    /**
     * Add light source trait.
     *
     * @param opts - Light source options
     * @returns this (for chaining)
     */
    lightSource(opts = {}) {
        this._lightSource = opts;
        return this;
    }
    /**
     * Add a custom trait to the entity.
     *
     * @param trait - Any ITrait instance
     * @returns this (for chaining)
     */
    addTrait(trait) {
        this._traits.push(trait);
        return this;
    }
    /**
     * Bypass validation for placement (e.g., placing in closed containers).
     * Internally wraps the world in an AuthorModel.
     *
     * @returns this (for chaining)
     */
    skipValidation() {
        this._skipValidation = true;
        return this;
    }
    /**
     * Create the object entity with configured traits.
     *
     * @returns The created IFEntity
     */
    build() {
        const entity = this.world.createEntity(this.name, 'object');
        entity.add(new world_model_1.IdentityTrait({
            name: this.name,
            description: this._description,
            aliases: this._aliases,
        }));
        if (this._scenery) {
            entity.add(new world_model_1.SceneryTrait());
        }
        if (this._lightSource) {
            entity.add(new world_model_1.LightSourceTrait(this._lightSource));
        }
        for (const trait of this._traits) {
            entity.add(trait);
        }
        if (this._location) {
            const mover = this._skipValidation
                ? new world_model_1.AuthorModel(this.world.getDataStore(), this.world)
                : this.world;
            mover.moveEntity(entity.id, this._location.id);
        }
        return entity;
    }
}
exports.ObjectBuilder = ObjectBuilder;
