// @sharpee/helpers/builders/room.js
"use strict";
/**
 * RoomBuilder — fluent builder for room entities.
 *
 * Public interface: description, aliases, dark, build.
 *
 * Owner context: @sharpee/helpers (ADR-140)
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RoomBuilder = void 0;
const world_model_1 = require("../../world-model/index.js");
/**
 * Fluent builder for creating room entities.
 *
 * @example
 * ```typescript
 * const kitchen = room('Kitchen')
 *   .description('A warm kitchen.')
 *   .build();
 * ```
 */
class RoomBuilder {
    world;
    name;
    _description;
    _aliases;
    _isDark = false;
    _traits = [];
    constructor(world, name) {
        this.world = world;
        this.name = name;
    }
    /**
     * Set the room description.
     *
     * @param desc - Room description text
     * @returns this (for chaining)
     */
    description(desc) {
        this._description = desc;
        return this;
    }
    /**
     * Add name aliases for the room.
     *
     * @param names - Alternative names
     * @returns this (for chaining)
     */
    aliases(...names) {
        this._aliases = names;
        return this;
    }
    /**
     * Add a custom trait to the entity.
     *
     * @param trait - Any ITrait instance
     * @returns this (for chaining)
     */
    addTrait(trait) {
        this._traits.push(trait);
        return this;
    }
    /**
     * Mark the room as dark (requires a light source to see).
     *
     * @returns this (for chaining)
     */
    dark() {
        this._isDark = true;
        return this;
    }
    /**
     * Create the room entity with configured traits.
     *
     * @returns The created IFEntity
     */
    build() {
        const entity = this.world.createEntity(this.name, 'room');
        entity.add(new world_model_1.RoomTrait({ isDark: this._isDark }));
        entity.add(new world_model_1.IdentityTrait({
            name: this.name,
            description: this._description,
            aliases: this._aliases,
        }));
        for (const trait of this._traits) {
            entity.add(trait);
        }
        return entity;
    }
}
exports.RoomBuilder = RoomBuilder;
